rfcs/0018-glides-trainsheet-events.md

```diff
@@ -0,0 +1,571 @@
+
+# Drawbacks
+- This requires Glides to include new functionality to write their messages into a Kinesis stream. However, they already use `ex_aws` so additional messages should not be hard to add.
+ 
+# Rationale and alternatives
+- RTR already consumes a Kinesis stream of CloudEvents from OCS, 
+- so it's a limited additional lift to consume a new Kinesis stream
+- Kinesis Firehose allows writing the stream of messages to S3 for LAMP to consume, without needing a separate integration
+
+## Alternative: departure time as RFC3999 string
+This would be easy to parse with standard tools, but doesn't match existing OCS behavior which only provides a time.
+
+## Alternative: `dropped` boolean in Trip updated
+While this does eliminate the need for the Trip dropped and Trip restored messages, it also allows for confusing Trip updated messages. How should clients interpret a Trip updated message which included `dropped` True, along with other changes to the trip?
+
+## Alternative: split update types into different events
+Examples:
+- set departure time
+- set/unset operator
+- set/unset consist
+
+The current trainsheet interface does not distinguish between these cases: all changes are saved at the same time. Functionality to separate them would need to be added to Glides, or multiple events would need to be written in a batch, increasing duplication.
+
+## Alternative: GTFS-RT ServiceChange feed
+For much of CTD's history, we have used versions of GTFS-RT as our interchange format between applications. Current examples include:
+- Operator assignments and block waivers (Busloc -> Skate)
+- Arrival predictions at skipped stops (RTR -> Realtime Signs)
+
+If this data can be directly presented to riders, this avoids a processing step.
+
+However, this is also limiting:
+- both sides must already understand GTFS
+- can be hard to reproduce user intent
+- expects full (or at least mostly full) records
+
+While the ladder view in Glides understands GTFS-RT, the trainsheet information comes directly from HASTUS on a quarterly basis. It does not have updated trip IDs, so RTR would need to process the data anyways: it could not be provided directly to riders.
+
+# Unresolved questions
+- While some sample CloudEvent formats are included here, the specific format should be settled on as a conversation between the Glides and Transit Data teams.
+- The UI does not currently give a `reason` for edits, only for dropped trips. Should there be a `reason` for edits (separate from `comment`)?
+- Some trainsheet entries (removing an operator from their trips) can happen the day before. Do we need to persist any events longer than the 24 hour Kinesis default?
+- If Glides starts being used at non-terminal, non-yard stations, trip departure times may not be unique. (Government Center)
```

Comment:
I don't think this matters now that we are no longer assuming that trip departure times are unique.